src/test/java/RecipeTest.java

```diff
@@ -191,6 +193,63 @@ public void removeStepFailureTooLow() {
     recipe.removeStep(-10);
   }
 
+  @Test
+  public void addTag() {
+    Set<String> expected = new HashSet<>(Arrays.asList("grilledcheese", "quick", "cheese"));
+    recipe.addTag("cheese");
+    Assert.assertEquals(expected, recipe.getTags());
+  }
+
+  @Test
+  public void removeTag() {
+    Set<String> expected = new HashSet<>(Arrays.asList("grilledcheese"));
+    recipe.removeTag("quick");
+    Assert.assertEquals(expected, recipe.getTags());
+  }
+
+  @Test
+  public void removeNonexistentTag() {
+    Set<String> expected = new HashSet<>(Arrays.asList("grilledcheese", "quick"));
+    recipe.removeTag("chicken");
+    Assert.assertEquals(expected, recipe.getTags());
+  }
+
+  @Test
+  public void addIngredient() {
+    Set<String> expected = new HashSet<>(Arrays.asList("bread", "cheese", "butter", "oil"));
+    recipe.addIngredient("oil");
+    Assert.assertEquals(expected, recipe.getIngredients());
+  }
+
+  @Test
+  public void removeIngredient() {
+    Set<String> expected = new HashSet<>(Arrays.asList("cheese", "butter"));
+    recipe.removeIngredient("bread");
+    Assert.assertEquals(expected, recipe.getIngredients());
+  }
+
+  @Test
+  public void removeNonexistentIngredient() {
+    Set<String> expected = new HashSet<>(Arrays.asList("bread", "cheese", "butter"));
+    recipe.removeIngredient("chicken");
+    Assert.assertEquals(expected, recipe.getIngredients());
+  }
+
+  @Test
+  public void addSpinOff() {
+    SpinOff spinOff = new SpinOff(recipe);
+    Set<SpinOff> expected = new HashSet<>(Arrays.asList(spinOff));
+    recipe.addSpinOff(spinOff);
+    Assert.assertEquals(expected, recipe.getSpinOffs());
+  }
+
+  @Test
+  public void removeSpinOff() {
+    Set<SpinOff> expected = new HashSet<SpinOff>();
```

Comment:
Does this work? You no longer have the global SPINOFF.

Reply:
It definitely does not, my mistake